src/common/bitfield_util.h

```diff
@@ -0,0 +1,180 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <cstring>
+#include <limits>
+#include <stdexcept>
+#include <string>
+
+enum class BitfieldOverflowBehavior : uint8_t { kWrap, kSat, kFail };
+
+class BitfieldBitsLengthNotFitExecption : public std::runtime_error {
+ public:
+  BitfieldBitsLengthNotFitExecption() : std::runtime_error("bits length is not fitted to bitfield.") {}
+};
+
+class BitfieldEncoding {
+ public:
+  enum class Type : uint8_t { kSigned, kUnsigned };
+
+  // check whether the bit length is fit to given number sign.
+  // Redis has bits length limitation to bitfield.
+  // Quote:
+  // The supported encodings are up to 64 bits for signed integers, and up to 63 bits for unsigned integers. This
+  // limitation with unsigned integers is due to the fact that currently the Redis protocol is unable to return 64 bit
+  // unsigned integers as replies.
+  // see also https://redis.io/commands/bitfield/
+  static bool IsSupportedBitLengths(Type type, uint8_t bits) {
+    uint8_t max_bits = 64;
+    if (type == Type::kUnsigned) {
+      max_bits = 63;
+    }
+    return 1 <= bits && bits <= max_bits;
+  }
+
+  BitfieldEncoding(Type type, uint8_t bits) : type_(type), bits_(bits) {
```

Comment:
This can be wrapped as a `Result<BitfieldEncoding>` builder, and `BitfieldEncoding` as a private ctor.